gateway-js/src/loadCsdlFromStorage.ts

```diff
@@ -0,0 +1,100 @@
+import { fetch } from 'apollo-server-env';
+import { GraphQLError } from 'graphql';
+import { CsdlQuery } from './__generated__/graphqlTypes';
+
+export const CSDL_QUERY = /* GraphQL */`#graphql
+  query Csdl($apiKey: String!, $ref: String!) {
+    routerConfig(ref: $ref, apiKey: $apiKey) {
+      __typename
+      ... on RouterConfigResult {
+        id
+        csdl
+      }
+      ... on FetchError {
+        code
+        message
+      }
+    }
+  }
+`;
+
+type CsdlQueryResult = CsdlQuerySuccess | CsdlQueryFailure;
+
+interface CsdlQuerySuccess {
+  data: CsdlQuery;
+}
+
+interface CsdlQueryFailure {
+  data?: CsdlQuery;
+  errors: GraphQLError[];
+}
+
+const { name, version } = require('../package.json');
+
+export async function loadCsdlFromStorage({
+  graphId,
+  graphVariant,
+  apiKey,
+  endpoint,
+  fetcher,
+}: {
+  graphId: string;
+  graphVariant: string;
+  apiKey: string;
+  endpoint: string;
+  fetcher: typeof fetch;
+}) {
+  const result = await fetcher(endpoint, {
+    method: 'POST',
+    body: JSON.stringify({
+      query: CSDL_QUERY,
+      variables: {
+        ref: `${graphId}@${graphVariant}`,
+        apiKey,
+      },
+    }),
+    headers: {
+      'apollographql-client-name': name,
+      'apollographql-client-version': version,
+      'user-agent': `${name}/${version}`,
+      'content-type': 'application/json',
+    },
+  });
+
+  let response: CsdlQueryResult;
+
+  if (result.ok || result.status === 400) {
+    try {
+      response = await result.json();
+    } catch (e) {
+      // Bad response
+      throw new Error(result.status + ': ' + e.message ?? e);
+    }
+
+    // This happens before the 200 check below because the server returns a 400
```

Comment:
This comment is no longer correct (or needed)

Reply:
29d255e